content/utils.go

```diff
@@ -0,0 +1,65 @@
+/*
+Copyright The ORAS Authors.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package content
+
+import (
+	"errors"
+	"fmt"
+	"io"
+
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+	"oras.land/oras-go/v2/internal/ioutil"
+)
+
+var (
+	// ErrInvalidDescriptorSize is returned by ReadAll() when
+	// the descriptor has an invalid size.
+	ErrInvalidDescriptorSize = errors.New("invalid descriptor size")
+
+	// ErrMismatchedDigest is returned by ReadAll() when
+	// the descriptor has an invalid digest.
+	ErrMismatchedDigest = errors.New("mismatched digest")
+
+	// ErrTrailingData means that there exists trailing data unread
+	// when the read operation terminates.
+	ErrTrailingData = errors.New("trailing data")
+)
+
+// ReadAll safely reads the content described by the descriptor.
+// The read content is verified against the size and the digest.
+func ReadAll(r io.Reader, desc ocispec.Descriptor) ([]byte, error) {
+	if desc.Size < 0 {
+		return nil, ErrInvalidDescriptorSize
+	}
+	buf := make([]byte, desc.Size)
+
+	// verify while reading
+	verifier := desc.Digest.Verifier()
+	r = io.TeeReader(r, verifier)
+	_, err := io.ReadFull(r, buf)
+	// verify the size of the read content
```

Comment:
Does the comment relate to the next line code?

Reply:
Deleted this line of comment to avoid confusions.

Reply:
i see, I feel it might be better to move it above `_, err := io.ReadFull(r, buf)` and we can put it with err check together in the if statement.

Reply:
Changed accordingly.